serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt

```diff
@@ -50,57 +90,45 @@ class MatrixSession(val client: HttpClient, val access_token: String, var transa
     }
 
     fun closeSession() {
-        //Update Database with latest transactionId
+        // Update Database with latest transactionId
         Database.updateSession(this.access_token, this.transactionId)
         // TO ACT LIKE A LOGOUT, CLOSING THE CLIENT
         client.close()
+        sync_should_run = false
     }
-    fun sync(): Outcome<Unit> {
-        val timeout_ms = 10000
-        val url = if (sync_response != null) {
-            "https://synapse.room409.xyz/_matrix/client/r0/sync?since=${sync_response!!.next_batch}&timeout=$timeout_ms&access_token=$access_token"
-        } else {
-            val limit = 5
-            "https://synapse.room409.xyz/_matrix/client/r0/sync?filter={\"room\":{\"timeline\":{\"limit\":$limit}}}&access_token=$access_token"
-        }
-        try {
-            runBlocking {
-                val new_sync_response = client.get<SyncResponse>(url)
-                if (sync_response != null) {
-                    println("Sync response wasn't null")
-                    for ((room_id, room) in new_sync_response.rooms.join) {
-                        if (sync_response!!.rooms.join.containsKey(room_id)) {
-                            // This should actually be updated with messages from the timeline too
-                            sync_response!!.rooms.join[room_id]!!.state = room.state
-                            // This also needs to be changed to something that supports discontinuous
-                            // sections of timeline with different prev_patch, etc
-                            sync_response!!.rooms.join[room_id]!!.timeline.events += room.timeline.events
-                        } else {
-                            println(" adding in new room_id and room")
-                            sync_response!!.rooms.join[room_id] = room
-                        }
+    fun mergeInSync(new_sync_response: SyncResponse) {
+        synchronized(this) {
+            if (sync_response == null) {
+                sync_response = new_sync_response
+            } else {
+                for ((room_id, room) in new_sync_response.rooms.join) {
+                    if (sync_response!!.rooms.join.containsKey(room_id)) {
+                        // A little bit hacky, we just add all our new state events here, prepending
+                        // the new ones
```

Comment:
Why are the new state events being prepended instead of added to the end?

Reply:
Hehe, because I search through them from the front when looking for the most up-to-date information ;)